src/btree/bt_delete.c

```diff
@@ -372,21 +381,28 @@ __wt_delete_page_skip(WT_SESSION_IMPL *session, WT_REF *ref, bool visible_all)
     if (!WT_REF_CAS_STATE(session, ref, WT_REF_DELETED, WT_REF_LOCKED))
         return (false);
 
-    skip = !__wt_page_del_active(session, ref, visible_all);
+    /*
+     * Check visibility.
+     *
+     * Use the option to hide prepared transactions in all checks; we can't skip a page if the
+     * deletion is only prepared (we need to visit it to generate a prepare conflict), and we can't
+     * discard the page_del info either, as doing so leads to dropping the on-disk page and if the
+     * prepared transaction rolls back we'd then be in trouble.
+     */
+    if (visible_all)
+        skip = discard = __wt_page_del_visible_all(session, ref->page_del, true);
+    else {
+        skip = __wt_page_del_visible(session, ref->page_del, true);
+        discard = skip ? __wt_page_del_visible_all(session, ref->page_del, true) : false;
+    }
 
     /*
      * The fast-truncate structure can be freed as soon as the delete is stable: it is only read
```

Comment:
stable is misleading here. I think this means globally visible?

Reply:
Indeed. Fixed in [5431108](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/5431108959e98cdf9b3aa1b0fb0aa532b7ab0db8)